http_server_concurrent.py

```diff
@@ -0,0 +1,36 @@
+from echo_server import echo_server
+from echo_client import echo_client
+from http1 import http1, response_ok, response_error, parse_request
+from http2 import HttpServer
+from http2 import ResourceNotFound
+from gevent.server import StreamServer
+from gevent.monkey import patch_all
+from gevent.server import serve_forever
+
+
+# adapted from https://github.com/gevent/gevent/blob/master/examples/echoserver.py
+
+
+def handler(socket, address):
+    print 'New Connection'
+    buffsize = 32
+    try:
+        msg = ""
+        finish = False
+        while not finish:
+            in_msg = socket.recv(buffsize)
+            if in_msg:
+                msg += in_msg
+            else:
+               socket.sendall(parse_request(msg))
+               socket.close()
+               finish = True
+    except KeyboardInterrupt:
+        socket.close()
```

Comment:
You are not responsible for catching `KeyboardInterrupt` when using a `StreamServer`.  It is responsible for managing the existence of the server and closing the `socket` it gives you in case of unexpected termination.